File: src/Aiko.Core/SessionReminder.cs
using System.Text.Json.Nodes;

namespace Aiko.Core;

/// The line Claude Code shows at the start of a session when a command overrode a folder binding.
///
/// An explicit command wins over a binding, and the person may mean it (D-159). But limits are
/// money, so Claude Code says whose limits are being spent. The shim leaves AIKO_LAUNCH and
/// AIKO_ENVIRONMENT for the hook to read.
///
/// These two sentences live here and not in tools/strings.json: the bridge prints them, and the
/// bridge has no resources of the app. The copy was agreed on the artifact (D-163).
public static class SessionReminder
{
    public const string Argument = "--session-start";
    public const string LaunchVariable = "AIKO_LAUNCH";
    public const string EnvironmentVariable = "AIKO_ENVIRONMENT";
    public const string LaunchedByCommand = "command";

    public static string? MessageFor(
        string? launch,
        string? runningEnvironment,
        string workingDirectory,
        EnvironmentSettings settings,
        bool russian)
    {
        if (launch != LaunchedByCommand || string.IsNullOrWhiteSpace(runningEnvironment))
        {
            return null;
        }

        var bound = ProjectBinding.BoundEnvironmentFor(workingDirectory, settings);
        if (bound is null || bound.Name.Equals(runningEnvironment, StringComparison.Ordinal))
        {
            return null;
        }

        return russian
            ? $"Aiko: эта папка относится к {bound.Name}, а сессия запущена в {runningEnvironment}. Лимиты тратятся из {runningEnvironment}."
            : $"Aiko: this folder belongs to {bound.Name}, but this session runs in {runningEnvironment}, so the limits of {runningEnvironment} are used.";
    }

    /// The hook answer Claude Code reads. systemMessage is shown to the person; plain output of a
    /// SessionStart hook would go to the model instead.
    public static string HookOutput(string message) =>
        new JsonObject { ["systemMessage"] = message }.ToJsonString();

    /// Russian when chosen in settings, or when settings follow Windows and Windows speaks Russian.
    /// The bridge gets the language id of Windows, because it runs with invariant globalization.
    public static bool IsRussian(AikoLanguage language, int windowsUiLanguageId) =>
        language == AikoLanguage.Russian
        || (language == AikoLanguage.System && (windowsUiLanguageId & 0x3FF) == 0x19);

    /// The working folder from the JSON Claude Code gives a hook on stdin.
    public static string? WorkingDirectoryIn(string hookInput)
    {
        try
        {
            return JsonNode.Parse(hookInput) is JsonObject root
                && root.TryGetPropertyValue("cwd", out var cwd)
                && cwd is JsonValue value
                && value.TryGetValue<string>(out var text)
                    ? text
                    : null;
        }
        catch (System.Text.Json.JsonException)
        {
            return null;
        }
    }
}

File: mac/Sources/AikoKit/SessionReminder.swift
import Foundation

/// The line Claude Code shows at the start of a session when a command overrode a folder binding.
///
/// An explicit command wins over a binding, and the person may mean it (D-159). But limits are
/// money, so Claude Code says whose limits are being spent. The shim leaves AIKO_LAUNCH and
/// AIKO_ENVIRONMENT for the hook to read.
///
/// These two sentences live here and not in tools/strings.json: the bridge prints them, and the
/// bridge has no resources of the app. The copy was agreed on the artifact (D-163).
public enum SessionReminder {
    public static let argument = "--session-start"
    public static let launchVariable = "AIKO_LAUNCH"
    public static let environmentVariable = "AIKO_ENVIRONMENT"
    public static let launchedByCommand = "command"

    public static func messageFor(
        _ platform: PlatformConventions,
        launch: String?,
        runningEnvironment: String?,
        workingDirectory: String,
        settings: EnvironmentSettings,
        russian: Bool
    ) -> String? {
        guard launch == launchedByCommand,
              let running = runningEnvironment,
              !running.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return nil
        }

        let found = ProjectBinding.boundEnvironmentFor(
            platform, workingDirectory: workingDirectory, settings: settings)
        guard let bound = found?.name, bound != running else {
            return nil
        }

        return russian
            ? "Aiko: эта папка относится к \(bound), а сессия запущена в \(running). Лимиты тратятся из \(running)."
            : "Aiko: this folder belongs to \(bound), but this session runs in \(running), so the limits of \(running) are used."
    }

    /// The hook answer Claude Code reads. systemMessage is shown to the person; plain output of a
    /// SessionStart hook would go to the model instead.
    public static func hookOutput(_ message: String) -> String {
        JsonNode.object(JsonObject([("systemMessage", .string(message))])).toJsonString()
    }

    /// Russian when chosen in settings, or when settings follow the system and the system speaks
    /// Russian. The Windows bridge gets a language id, because it runs with invariant globalization.
    public static func isRussian(_ language: AikoLanguage, _ systemLanguageId: Int) -> Bool {
        language == .russian || (language == .system && (systemLanguageId & 0x3FF) == 0x19)
    }

    /// The same answer on macOS, where the system says its language as a tag such as "ru-RU".
    /// Windows hands out a number instead, so the two bridges ask this in their own way.
    public static func isRussian(_ language: AikoLanguage, systemLanguageTag: String?) -> Bool {
        if language == .russian {
            return true
        }

        guard language == .system, let tag = systemLanguageTag else {
            return false
        }

        let base = tag.prefix { $0 != "-" && $0 != "_" }
        return base.caseInsensitiveCompare("ru") == .orderedSame
    }

    /// The working folder from the JSON Claude Code gives a hook on stdin.
    public static func workingDirectoryIn(_ hookInput: String) -> String? {
        JsonNode.parse(hookInput)?.objectValue?["cwd"]?.stringValue
    }
}
